### PR TITLE
Simplify client instantiation

### PR DESCRIPTION
**Pull request by davidstanley01**
This PR simplifies the way we instantiate the client. To do this, I've swapped the HTTPlug stuff for just using guzzle. Along the way, I ported the plugins to be handlers in guzzle and also updated the tests and stuff to match. I've added integration with CodeClimate and test coverage provide by Travis (via CodeClimate). Finally, I've added a retry handler to the client with exponential backoff. All good.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,12 +47,69 @@ require_once 'vendor/autoload.php';
...
+$client->setAccessToken('abc123');
+
+// Set Client credentials if using that grant type
+$client->useClientCredentials($appId, $appSecret);
+
+// Use legacy clientId and apiKey
+$client->useLegacyToken($clientId, $apiKey);
+
+// Use a refresh token to get a new access token
+$client->useRefreshToken($appId, $appSecret, $refreshToken);
```
Comment (bkuhl) [inline README.md]: I _really_ like this approach, you've done a killer job with it!

File: src/Http/Authenticator.php [hunk 2]
```diff
@@ -4,29 +4,208 @@
...
+    /**
+     * @param string $appId
+     * @param string $appSecret
+     * @param string $refreshToken
+     */
+    public function useRefreshToken(string $appId, string $appSecret, string $refreshToken): void
+    {
+        $auth = new RefreshCredentials($appId, $appSecret, $refreshToken);
+
+        $this->auth = $auth;
```
Comment (bkuhl) [inline src/Http/Authenticator.php]: Can we skip the `$auth` and just set it direct via `$this->auth = new Refresh...`?

File: README.md (hunk 1 above)
Comment (davidstanley01) [reply README.md]: Thanks

File: src/Http/Authenticator.php (hunk 2 above)
Comment (davidstanley01) [reply src/Http/Authenticator.php]: done

**Comment by davidstanley01** [comment]
I've added a blurb to the readme about how to use the `mock($endpoint)` method on the `ApiClient` class.

File: README.md [hunk 3]
```diff
@@ -735,3 +793,26 @@ $lastUsers = $users->getLastPage();
...
+
+    $this->assertInstanceOf(WorkflowsEndpoint::class, $mockedWorkflows);
+    $this->assertInstanceOf(MockInterface::class, $mockedWorkflows);
+
+    $this->assertSame(
+        $mockedWorkflows,
+        $client->workflows()
+    );
+}
+```
```
Comment (bkuhl) [inline README.md]: What do you think about replacing this with an example of how the mock would be used within a phpunit test?

Comment (davidstanley01) [reply README.md]: that is how it would be used...

**Pull request merged**